### ISSUE TITLE
environment (e.g. local, staging, prod) specific config

### ISSUE THREAD

**Issue by nmaier95** [fictionco/fiction#136]
**Reason for this issue**
Unsure on how this topic is supposed to work with factor. In the "Before you start" section of your documentation we are introduced how and where to store environment-variables.
What about the case that those change from e.g. local to staging to production?
How is one supposed to deal with this issue? How can i tell factor to use a different config file?

**Comment by arpowers** [comment]
@nmaier95 

that changes depending on the way you run your app, for example, with Heroku (what we use) they have a settings interface for the app where you add each env variable (see screenshot)... 

<img width="781" alt="Screen Shot 2020-02-18 at 4 03 11 PM" src="https://user-images.githubusercontent.com/308134/74785719-348db500-5268-11ea-9e15-65a7bf6b8076.png">

@nodesocket can you clarify how you set this up in your setup? 

Another thing you could try is looking at how [dotenv](https://www.npmjs.com/package/dotenv) is handled in your chosen environment. (e.g. google "digital ocean node dotenv")

I'll clarify the docs also.

**Comment by nmaier95** [comment]
Hi @arpowers thank you very much for your quick response. I'll have a look at that.
I also just discovered another issue with the .env file not being commited. Unfortunately docker-compose uses this (and only this) .env file to store e.g. PROJECT_NAME or COMPOSE_FILE. Those configs belong to any developer on a project and are therefore needed to be checked in imo. Also docker-compose has not come up yet with any kind of configureable parameter to overwrite the default .env regarding those properties mentioned before.

How shall we continue with this?

**Comment by arpowers** [comment]
you NEVER want your `.env` file in source control. 

The "best practice" is to set up your environment variables manually for each environment.  This file is for secret information; anything that doesn't need to be secret can be in config/settings and stored in source control.

As a story, we once committed our SMTP password to source control accidentally; it was hacked and used as a spam account within 3 hours.

**Comment by nmaier95** [comment]
Im aware of the fact that .env is not supposed to be in source control. However I haven't figured out how else to set those two docker-compose settings. Also those do not hurt anyone if they are exposed. To be fair though, I never stored something else inside of this file yet.

**Comment by arpowers** [comment]
@nodesocket can you help? you're the docker master.

**Comment by nodesocket** [comment]
@nmaier95 where is the `docker-compose.yaml` file you are using?

Typically, this is handled by the platform you are deploying to. I.E. Heroku has environment variables you specify via their web interface.

When I build my Factor Docker container I am using the following Dockerfile _(I am not using compose)_:

```
FROM node:12.14.1-buster-slim

RUN apt-get update && apt-get -y install --no-install-recommends \
unzip \
bzip2 \
ca-certificates \
curl && \
apt-get -y autoremove && \
apt-get clean && \
rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*

RUN mkdir -p /var/factor
WORKDIR /var/factor

COPY . .

RUN yarn install --production
RUN yarn factor build
RUN rm -f .env

EXPOSE 3000

RUN chown -R node /var/factor
USER node
CMD ["yarn", "factor", "serve"]
```

.dockerignore

```
.DS_Store
.dockerignore
.env.example
.git
.gitignore
.nvmrc
Dockerfile
node_modules
npm-debug.log
package-lock.json
yarn.lock
yarn-error.log
```

Note, that I am copying my `.env` file into the container from local in order to run `yarn factor build`, but then I immediately delete it. **So this has the unfortunate site effect that the Docker container build is tied to the `.env` file.** I.E. don't think you can have a single Docker container for all environments unless you run `yarn factor build` on the container startup instead of in the build process.

**Comment by nmaier95** [comment]
Besides me using compose, our Dockerfiles look similar. But deleting the .env file after building factor is a good point!

i have an almost empty (no services, only version and volumes) compose file in my project root. Then I have under build/docker/ (path doesn't really matter) for each service its own compose.yml file. Those are then combined in the property of the .env file COMPOSE_FILE.

For now I switched to an .env.example which I added to version control. This example contains my compose configs and sample key-value pairs factor needs. My idea is that this sample file may then be used to generate an .env file inside of the container. Maybe even with auto generated tokens and so on.

Appreciate your guys help!
